Use smart pointers in WebFrameProxy

pratiksha-99 opened the pull request:
#### 05b9f6e30e2944947828755ff6d00931e3864637
<pre>
Use smart pointers in WebFrameProxy
<a href="https://bugs.webkit.org/show_bug.cgi?id=287208">https://bugs.webkit.org/show_bug.cgi?id=287208</a>

Reviewed by Chris Dumez.

Addressed the smart pointer static analyzer warnings.

* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::webProcessWillShutDown):
(WebKit::WebFrameProxy::bindAccessibilityFrameWithData):
(WebKit::WebFrameProxy::prepareForProvisionalLoadInProcess):
(WebKit::WebFrameProxy::getFrameInfo):
(WebKit::WebFrameProxy::webPageIDInCurrentProcess):
(WebKit::WebFrameProxy::traverseNext const):

* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::webProcessWillShutDown):
(WebKit::WebFrameProxy::bindAccessibilityFrameWithData):
(WebKit::WebFrameProxy::prepareForProvisionalLoadInProcess):
(WebKit::WebFrameProxy::getFrameInfo):
(WebKit::WebFrameProxy::webPageIDInCurrentProcess):
(WebKit::WebFrameProxy::traverseNext const):

Canonical link: <a href="https://commits.webkit.org/290054@main">https://commits.webkit.org/290054@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a7a53f86eba48a10b6ae77ca3ee545a686b4da24

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88824 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8348 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43291 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93795 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39583 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8735 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16532 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/68459 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26144 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91827 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/6674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80315 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48824 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/6429 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38691 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/76792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35629 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/95670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16004 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77329 "42 flakes 100 failures") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16260 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76174 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/76612 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18883 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20998 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/9069 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16018 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/21329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/15759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/19210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17540 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -150,9 +150,9 @@ void WebFrameProxy::webProcessWillShutDown()
-        m_activeListener = nullptr;
+    if (RefPtr activeListener = m_activeListener) {
+        activeListener->ignore();
+        activeListener = nullptr;
```
This is wrong. We want to null out the data member, not the local variable.

cdumez commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -211,7 +211,8 @@ void WebFrameProxy::bindAccessibilityFrameWithData(std::span<const uint8_t> data
         return;
 
-    m_page->sendToProcessContainingFrame(m_frameID, Messages::WebProcess::BindAccessibilityFrameWithData(m_frameID, data));
+    if (RefPtr page = m_page.get())
```
We already do a null check above, please drop the one above if you do it here.

cdumez commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -434,10 +435,12 @@ void WebFrameProxy::prepareForProvisionalLoadInProcess(WebProcessProxy& process,
-    page->websiteDataStore().protectedNetworkProcess()->addAllowedFirstPartyForCookies(process, mainFrameDomain, LoadedWebArchive::No, [pageID = page->webPageIDInProcess(process), completionHandler = WTFMove(completionHandler)] mutable {
-        completionHandler(pageID);
-    });
+    if (RefPtr<WebPreferences> preferences = adoptRef(page->preferences()))
```
Uh. Not sure what you're trying to do here but this cannot be correct.

pratiksha-99 replied:
Sorry, I misunderstood the conversion here. The warning for `page->preferences()`  is `Call argument is uncounted and unsafe`, what's the best way to convert this ?

cdumez replied:
usually call `page->protectedPreferences()`

cdumez commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -490,7 +493,10 @@ void WebFrameProxy::getFrameInfo(CompletionHandler<void(FrameTreeNodeData&&)>&&
     }, *webPageIDInCurrentProcess());
 
-    bool isSiteIsolationEnabled = page() && page()->preferences().siteIsolationEnabled();
+    RefPtr<WebPageProxy> safePage = page();
```
`RefPtr page = this->page()`

cdumez commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -566,7 +572,7 @@ void WebFrameProxy::notifyParentOfLoadCompletion(WebProcessProxy& childFrameProc
 {
     if (m_page)
-        return m_page->webPageIDInProcess(process());
+        return m_page->webPageIDInProcess(this->protectedProcess());
```
`this->` shouldn't be needed?

cdumez commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -599,8 +605,11 @@ auto WebFrameProxy::traverseNext(CanWrap canWrap) const -> TraversalResult
-        if (m_page)
-            return { m_page->protectedMainFrame(), DidWrap::Yes };
+        if (m_page) {
+            if (RefPtr page = m_page.get())
```
Again, double null check of the page.

cdumez commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -211,7 +211,8 @@ void WebFrameProxy::bindAccessibilityFrameWithData(std::span<const uint8_t> data
         return;
 
-    m_page->sendToProcessContainingFrame(m_frameID, Messages::WebProcess::BindAccessibilityFrameWithData(m_frameID, data));
+    RefPtr page = m_page.get();
```
Ideally we would do:
```
if (RefPtr page = m_page.get())
    page->sendToProcessContainingFrame(m_frameID, Messages::WebProcess::BindAccessibilityFrameWithData(m_frameID, data));
```
And drop the null check above:
```
    if (!m_page)
        return;
```

cdumez commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -566,7 +568,7 @@ void WebFrameProxy::notifyParentOfLoadCompletion(WebProcessProxy& childFrameProc
 {
     if (m_page)
-        return m_page->webPageIDInProcess(process());
+        return m_page->webPageIDInProcess(protectedProcess());
```
`webPageIDInProcess()` is not a trivial function so this should be:
```
if (RefPtr page = m_page.get())
    return page->webPageIDInProcess(protectedProcess());
```

cdumez commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -490,7 +488,8 @@ void WebFrameProxy::getFrameInfo(CompletionHandler<void(FrameTreeNodeData&&)>&&
 
-    bool isSiteIsolationEnabled = page() && page()->preferences().siteIsolationEnabled();
+    RefPtr page = this->page();
+    bool isSiteIsolationEnabled = page->preferences().siteIsolationEnabled();
```
We lost the null check here, you're missing `page && `

pratiksha-99 replied:
Fixed, I will be extra careful for these changes henceforth.

pratiksha-99 committed:
Use smart pointers in WebFrameProxy
https://bugs.webkit.org/show_bug.cgi?id=287208

Reviewed by Chris Dumez.

Addressed the smart pointer static analyzer warnings.

* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::webProcessWillShutDown):
(WebKit::WebFrameProxy::bindAccessibilityFrameWithData):
(WebKit::WebFrameProxy::prepareForProvisionalLoadInProcess):
(WebKit::WebFrameProxy::getFrameInfo):
(WebKit::WebFrameProxy::webPageIDInCurrentProcess):
(WebKit::WebFrameProxy::traverseNext const):

* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::webProcessWillShutDown):
(WebKit::WebFrameProxy::bindAccessibilityFrameWithData):
(WebKit::WebFrameProxy::prepareForProvisionalLoadInProcess):
(WebKit::WebFrameProxy::getFrameInfo):
(WebKit::WebFrameProxy::webPageIDInCurrentProcess):
(WebKit::WebFrameProxy::traverseNext const):

Canonical link: https://commits.webkit.org/290054@main
